tfhe/src/integer/server_key/radix_parallel/bit_extractor.rs

```diff
@@ -23,46 +50,125 @@ impl<'a> BitExtractor<'a> {
-        jobs.into_par_iter()
-            .map(|(block_index, bit_index)| {
-                let block = &blocks[block_index];
-                let lut = &self.bit_extract_luts[bit_index];
-                self.server_key.key.apply_lookup_table(block, lut)
-            })
-            .collect_into_vec(&mut bits);
+    pub(crate) fn prepare_next_batch(&mut self) {
+        if self.buffer.is_empty() {
+            let Some(next_block_to_extract_from) = self.input_blocks.next() else {
+                return;
+            };
+
+            match &self.bit_extract_luts {
+                None => self.buffer.push_back(next_block_to_extract_from.clone()),
+                Some(bit_extract_luts) => {
+                    let new_bits = self
+                        .server_key
+                        .key
+                        .apply_many_lookup_table(next_block_to_extract_from, bit_extract_luts);
+
+                    self.buffer.extend(new_bits);
+                }
+            }
+        }
+    }
+
+    pub(crate) fn prepare_all_bits(&mut self) {
+        let num_blocks = self.input_blocks.len();
+        self.prepare_n_bits(num_blocks * self.bits_per_block)
+    }
+
+    /// Extract the remaining `n` bits in parallel from the current source blocks
+    /// and place them into the internal buffer
+    ///
+    /// # Panics
+    ///
+    /// Panics if the current slice of blocks does not have less than n bits available
+    pub(crate) fn prepare_n_bits(&mut self, n: usize) {
+        if self.buffer.len() >= n {
+            return;
+        }
+
+        let num_bits_to_extract = n - self.buffer.len();
+        let num_blocks_to_process = num_bits_to_extract.div_ceil(self.bits_per_block);
+        let blocks = self.input_blocks.as_slice();
+
+        if let Some(bit_extract_luts) = &self.bit_extract_luts {
+            let mut new_bits = blocks[..num_blocks_to_process]
+                .par_iter()
+                .flat_map(|block| {
+                    self.server_key
+                        .key
+                        .apply_many_lookup_table(block, bit_extract_luts)
+                })
+                .collect::<Vec<_>>();
+            self.buffer.extend(new_bits.drain(..));
+        } else {
+            let iterator = blocks[..num_blocks_to_process].iter().cloned();
+            self.buffer.extend(iterator);
+        };
+
+        // We have to advance our internal iterator
+        self.input_blocks = blocks[num_blocks_to_process..].iter();
+    }
+
+    /// Extract all the remaining bits in parallel from the current source blocks
+    pub(crate) fn extract_all_bits(&mut self) -> Vec<Ciphertext> {
+        let num_blocks = self.input_blocks.len();
+        self.extract_n_bits(num_blocks * self.bits_per_block)
+    }
 
+    /// Extract the remaining `n` bits in parallel from the current source blocks
+    ///
+    /// # Panics
+    ///
+    /// Panics if the current slice of blocks does not have less than n bits available
```

Comment:
`/// Panics if the current slice of blocks have less than n bits available`
or
`/// Panics if the current slice of blocks does not have n bits available`